### PR TITLE
Adjust constraints check for reduced extension methods to avoid failures for type parameters that couldn't be inferred from the first argument.

### PR DESCRIPTION
**Pull request by AlekseyTs**
@VSadov, @gafter, @agocke, @jaredpar Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs [hunk 1]
```diff
@@ -73,12 +73,35 @@ public static MethodSymbol InferExtensionMethodTypeArguments(this MethodSymbol m
...
+            }
+
             // Check constraints.
             var diagnosticsBuilder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
             var typeParams = method.TypeParameters;
-            var substitution = new TypeMap(typeParams, typeArgs);
+            var substitution = new TypeMap(typeParams, typeArgsForConstraintsCheck);
             ArrayBuilder<TypeParameterDiagnosticInfo> useSiteDiagnosticsBuilder = null;
-            var success = method.CheckConstraints(conversions, substitution, method.TypeParameters, typeArgs, compilation, diagnosticsBuilder, ref useSiteDiagnosticsBuilder);
+            var success = method.CheckConstraints(conversions, substitution, typeParams, typeArgsForConstraintsCheck, compilation, diagnosticsBuilder, ref useSiteDiagnosticsBuilder);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs]: success appears to be ignored. Should a constraint failure cause the method to be rejected?

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs [hunk 2]
```diff
@@ -3615,5 +3615,54 @@ public static void Foo(this int x)
                 //     using X = N.S;
                 Diagnostic(ErrorCode.HDN_UnusedUsingDirective, "using X = N.S;").WithLocation(4, 5));
         }
+
+        [WorkItem(1094849, "DevDiv")]
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs]: Please port issues to github when possible (scrubbing private data if necessary) so others can make as much sense of pull requests as we can.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs]: success is checked on line 120, GitHub hides it.

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs]: Opened #2288

**Comment by AlekseyTs** [comment]
@gafter I responded to your's comments

**Comment by gafter** [comment]
LGTM

**Comment by VSadov** [comment]
LGTM

**Pull request merged**